### PR TITLE
[webgui] let use unix sockets, introduce `rootssh.sh` script

### PR DESCRIPTION
**Pull request by linev**
1. Support unix sockets in `THttpServer`, fix civetweb warning
2. Let configure unix socket for webgui widgets
3. Introduce `rootssh.sh` script which use such ferature

In `rootssh.sh` special tunnel is created which forward requests from local http port to remote unix socket.
In remote session root will automatically use socket which specified as `ROOT_WEBGUI_SOCKET` environment variable.
This socket file by default get `0700` mode - means only user can accuse it. 

This solves most of security issues with using of ROOT webgui on public nodes!

`rootssh.sh`  also creates listener which get informed when new window is started on remote session
and automatically popup web browser using `xdg-open $url` command

Typical session should look like:
```
[localhost] rootssh.sh user@remotehost
[remotehost] root --web=server -e "new Browser"
```

### CODE REVIEW

File: config/rootssh.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+# start ssh
+
+ssh -t -R $listener_remote:$listener_local -L $localport:$root_socket $1 $2 $3 $4 $5 "$cmd"
+
+echo "stop" | netcat -U $listener_local -q 1
+
+echo "Kill listener process $listener_id"
+kill -9 $listener_id > /dev/null 2>&1
+
+echo "Remove temporary files"
```
Comment (Axel-Naumann) [inline config/rootssh.sh:86]: You might want to remove these files even upon <kbd>Ctrl<kbd></kbd>C</kbd>: https://stackoverflow.com/questions/12771909/bash-using-trap-ctrlc

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 2]
```diff
@@ -154,6 +155,41 @@ void RWebWindowsManager::AssignWindowThreadId(RWebWindow &win)
    }
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+/// If LC_ROOT_LISTENER_SOCKET variable is configured,
+/// message will be send to that unix socket
```
Comment (Axel-Naumann) [inline gui/webdisplay/src/RWebWindowsManager.cxx:160]:
```suggestion
/// If ROOT_LISTENER_SOCKET variable is configured,
/// message will be sent to that unix socket
```

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 3]
```diff
@@ -154,6 +155,41 @@ void RWebWindowsManager::AssignWindowThreadId(RWebWindow &win)
...
+
+#else
+
+   const char *fname = gSystem->Getenv("ROOT_LISTENER_SOCKET");
+   if (!fname || !*fname)
+      return false;
+
+   TSocket s(fname);
+   if (!s.IsValid())
+      return false;
```
Comment (Axel-Naumann) [inline gui/webdisplay/src/RWebWindowsManager.cxx:176]: Should this be emitting a warning / error message?

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 4]
```diff
@@ -154,6 +155,41 @@ void RWebWindowsManager::AssignWindowThreadId(RWebWindow &win)
...
+
+   int res = s.SendRaw(msg.c_str(), msg.length());
+
+   s.Close();
+
+   if (res > 0) {
+      // workaround to let listener configure port forwarding
+      gSystem->ProcessEvents();
+      gSystem->Sleep(10);
```
Comment (Axel-Naumann) [inline gui/webdisplay/src/RWebWindowsManager.cxx:185]: Why 10 and not 2 or 42? :-) Any chance that we can have a hand-shake, i.e. a second message that conveys "listener is ready"?

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 5]
```diff
@@ -182,6 +218,14 @@ void RWebWindowsManager::AssignWindowThreadId(RWebWindow &win)
 ///      WebGui.UseHttps: yes
 ///      WebGui.ServerCert: sertificate_filename.pem
 ///
+/// Alternatively, one can specify unix socket to handle requests:
+///
+///      WebGui.UnixSocket: /path/to/unix/socket
+///      WebGui.UnixSocketMode: 0700
+///
+/// Typically one used unix sockets together with server mode like `root --web=server:/tmp/root.socket` and
+/// then redirect it via ssh tunel to client node
```
Comment (Axel-Naumann) [inline gui/webdisplay/src/RWebWindowsManager.cxx:227]:
```suggestion
/// then redirect it via ssh tunnel (e.g. using `rootssh`) to client node
```

File: cmake/modules/RootConfiguration.cmake [hunk 6]
```diff
@@ -854,6 +854,7 @@ configure_file(${CMAKE_SOURCE_DIR}/config/setxrd.sh ${CMAKE_RUNTIME_OUTPUT_DIREC
 configure_file(${CMAKE_SOURCE_DIR}/config/proofserv.in ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/proofserv @ONLY NEWLINE_STYLE UNIX)
 configure_file(${CMAKE_SOURCE_DIR}/config/roots.in ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/roots @ONLY NEWLINE_STYLE UNIX)
 configure_file(${CMAKE_SOURCE_DIR}/config/root-help.el.in root-help.el @ONLY NEWLINE_STYLE UNIX)
+configure_file(${CMAKE_SOURCE_DIR}/config/rootssh.sh ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/rootssh.sh @ONLY NEWLINE_STYLE UNIX)
```
Comment (Axel-Naumann) [inline cmake/modules/RootConfiguration.cmake:857]:
```suggestion
configure_file(${CMAKE_SOURCE_DIR}/config/rootssh.sh ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/rootssh @ONLY NEWLINE_STYLE UNIX)
```
We generally call "things" without ".py" or ".sh" extension - that's an implementation detail :-)

**Review by Axel-Naumann — APPROVED** [review]
LGTM but see comments

File: config/rootssh.sh (hunk 1 above)
Comment (linev) [reply config/rootssh.sh:86]: Files will be removed in all cases.
`Ctrl-C` will be pressed by user in the ssh session and catched there.
In all cases ssh will exit normally and this code is executed.
Special listener process does not have interactive shell - it just receives via socket `new window` messages

File: gui/webdisplay/src/RWebWindowsManager.cxx (hunk 3 above)
Comment (linev) [reply gui/webdisplay/src/RWebWindowsManager.cxx:176]: `ROOT_LISTENER_SOCKET` is nice addon, but not really important - just new window may not be handled.
But I can add warning that socket is gone or not accessible.

File: gui/webdisplay/src/RWebWindowsManager.cxx (hunk 4 above)
Comment (linev) [reply gui/webdisplay/src/RWebWindowsManager.cxx:185]: Seems to be, timeout value is not important. 
It is just `sleep` system call which is matter - it let handle some other system events and let process socket sending.
I can edit comment here - port forwarding is now configured at very beginning automatically. 

By the way, I tried handshaking, but socket configured only in one direction. I need to make another sockets pair, which is not necessary here.

File: cmake/modules/RootConfiguration.cmake (hunk 6 above)
Comment (linev) [reply cmake/modules/RootConfiguration.cmake:857]: Maybe rename `rootssh.sh` into `rootssh` already in sources?
Name of script also appears in `INSTALL` instructions.

And in general script can be used without ROOT on local machine.

File: gui/webdisplay/src/RWebWindowsManager.cxx (hunk 4 above)
Comment (Axel-Naumann) [reply gui/webdisplay/src/RWebWindowsManager.cxx:187]: OK thanks, all good then. (10 seconds is a lot of time, though.)

Comment (linev) [reply gui/webdisplay/src/RWebWindowsManager.cxx:187]: It is 10 ms.

File: cmake/modules/RootConfiguration.cmake (hunk 6 above)
Comment (linev) [reply cmake/modules/RootConfiguration.cmake:857]: I will change this name in next PR.

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 7]
```diff
@@ -28,6 +28,7 @@
 #include "TROOT.h"
 #include "TEnv.h"
 #include "TExec.h"
+#include "TSocket.h"
```
Comment (bellenot) [inline gui/webdisplay/src/RWebWindowsManager.cxx:31]: That makes the build failing on Windows. And this requires the following change:
```
diff --git a/gui/webdisplay/CMakeLists.txt b/gui/webdisplay/CMakeLists.txt
index 50542b3ff6..1f68a838dd 100644
--- a/gui/webdisplay/CMakeLists.txt
+++ b/gui/webdisplay/CMakeLists.txt
@@ -23,6 +23,7 @@ ROOT_STANDARD_LIBRARY_PACKAGE(ROOTWebDisplay
        Core
        RHTTP
        MathCore
+       Net
 )

 ROOT_ADD_TEST_SUBDIRECTORY(test)
```

File: gui/webdisplay/src/RWebWindowsManager.cxx [hunk 8]
```diff
@@ -154,6 +155,43 @@ void RWebWindowsManager::AssignWindowThreadId(RWebWindow &win)
    }
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+/// If ROOT_LISTENER_SOCKET variable is configured,
+/// message will be sent to that unix socket
+
+bool RWebWindowsManager::InformListener(const std::string &msg)
+{
+#ifdef R__WIN32
```
Comment (bellenot) [inline gui/webdisplay/src/RWebWindowsManager.cxx:164]: Why not Win32? `TSocket` works on Windows...

Comment (linev) [reply gui/webdisplay/src/RWebWindowsManager.cxx:164]: listener used only together with `rootssh`. 
In this case small messages are written to unix socket (not TCP) when web window is shown.
I have no idea if unix sockets are working on Windows?
And if one can use `ssh windows_node`.

File: gui/webdisplay/src/RWebWindowsManager.cxx (hunk 7 above)
Comment (linev) [reply gui/webdisplay/src/RWebWindowsManager.cxx:31]: Sorry, please see #11824

File: gui/webdisplay/src/RWebWindowsManager.cxx (hunk 8 above)
Comment (bellenot) [reply gui/webdisplay/src/RWebWindowsManager.cxx:164]: OK, thanks for the info. That would be interesting to try with ssh on Windows, but it's not high priority

**Pull request merged**